File: java/session/src/se/pitch/oss/fedpro/common/session/SequenceNumber.java
/*
 *  Copyright (C) 2022 Pitch Technologies AB
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package se.pitch.oss.fedpro.common.session;

// Because of its wrap around behavior, the sequence number will never be negative unless explicitly set to invalid
// sequence number.
// Not thread-safe.

import java.util.Objects;

public class SequenceNumber {

   /*
      NO_SEQUENCE_NUMBER:
      Binary:     10000000 0000000 00000000 00000000
      Hex:        80 00 00 00
      Decimal:    2 147 483 648
      java int:  -2 147 483 648     (=Integer.MIN_VALUE)

      Highest possible sequence number:
      Binary:     01111111 11111111 11111111 11111111
      Hex:        7F FF FF FF
      Decimal:    2 147 483 647
      java int:   2 147 483 647     (=Integer.MAX_VALUE)
   */
   public static final int NO_SEQUENCE_NUMBER = Integer.MIN_VALUE;
   public static final int INITIAL_SEQUENCE_NUMBER = 0;
   public static final int MAX_SEQUENCE_NUMBER = Integer.MAX_VALUE;

   private int _value;

   public SequenceNumber(int value)
   {
      _value = value;
   }

   public SequenceNumber(SequenceNumber other)
   {
      this(other._value);
   }

   public SequenceNumber increment()
   {
      _value = resetInvalidToInitialValue(_value + 1);
      return this;
   }

   public static int nextAfter(int sequenceNumber) {
      return resetInvalidToInitialValue(sequenceNumber + 1);
   }

   public static int sum(int a, int b)
   {
      return wrapIntoValidRange(a + b);
   }

   public void set(SequenceNumber newNumber)
   {
      set(newNumber.get());
   }

   public SequenceNumber set(int newValue)
   {
      _value = newValue;
      return this;
   }

   public int get()
   {
      return _value;
   }

   private static int wrapIntoValidRange(int value)
   {
      // Sets the most significant bit to 0.
      return value & ~(1 << 31);
   }

   private static int resetInvalidToInitialValue(int value)
   {
      return Math.max(value, INITIAL_SEQUENCE_NUMBER);
   }

   protected static void validateArguments(int... values)
   {
      for (int value : values) {
         if (!isValidAsSequenceNumber(value)) {
            throw new IllegalArgumentException("Sequence number argument must be positive");
         }
      }
   }

   public static boolean isValidAsSequenceNumber(int value)
   {
      return value > -1;
   }

   public static boolean inInterval(
         int candidate,
         int oldest,
         int newest)
   {
      validateArguments(oldest, candidate, newest);
      if (oldest <= newest) {
         // When history is linear.
         return oldest <= candidate && candidate <= newest;
      } else {
         // When wrap around happens in history.
         return candidate <= newest || oldest <= candidate;
      }
   }

   public static boolean notInInterval(
         int candidate,
         int oldest,
         int newest)
   {
      return !inInterval(candidate, oldest, newest);
   }

   @Override
   public String toString()
   {
      return String.valueOf(_value);
   }

   @Override
   public boolean equals(Object o)
   {
      if (this == o) {
         return true;
      }
      if (o == null || getClass() != o.getClass()) {
         return false;
      }
      SequenceNumber that = (SequenceNumber) o;
      return _value == that._value;
   }

   @Override
   public int hashCode()
   {
      return Objects.hash(_value);
   }
}

File: cpp/src/fedpro/session/SequenceNumber.cpp
/***********************************************************************
  Copyright (C) 2023 Pitch Technologies AB

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 **********************************************************************/

#include "SequenceNumber.h"

#include <iostream>
#include <stdexcept>

namespace FedPro
{
   const int32_t SequenceNumber::NO_SEQUENCE_NUMBER;
   const int32_t SequenceNumber::INITIAL_SEQUENCE_NUMBER;
   const int32_t SequenceNumber::MAX_SEQUENCE_NUMBER;

   SequenceNumber::SequenceNumber(int32_t value)
   {
      _value = value;
   }

   SequenceNumber & SequenceNumber::increment() noexcept
   {
      if (_value == MAX_SEQUENCE_NUMBER || !isValidAsSequenceNumber(_value)) {
         _value = INITIAL_SEQUENCE_NUMBER;
      } else {
         _value++;
      }
      return *this;
   }

   void SequenceNumber::set(const SequenceNumber & newNumber)
   {
      set(newNumber.get());
   }

   void SequenceNumber::set(int32_t newValue)
   {
      _value = newValue;
   }

   int32_t SequenceNumber::get() const noexcept
   {
      return _value;
   }

   bool SequenceNumber::isValidAsSequenceNumber(int32_t value) noexcept
   {
      return value > -1;
   }

   bool SequenceNumber::inInterval(
         int32_t candidate,
         int32_t oldest,
         int32_t newest) noexcept
   {
      validateArguments({candidate, oldest, newest});
      if (oldest <= newest) {
         // When history is linear.
         return oldest <= candidate && candidate <= newest;
      } else {
         // When wrap around happens in history.
         return candidate <= newest || oldest <= candidate;
      }
   }

   bool SequenceNumber::notInInterval(
         int32_t candidate,
         int32_t oldest,
         int32_t newest) noexcept
   {
      return !inInterval(candidate, oldest, newest);
   }

   std::string SequenceNumber::toString() const noexcept
   {
      return std::to_string(_value);
   }

   int32_t SequenceNumber::nextAfter(int32_t number)
   {
      if (number == MAX_SEQUENCE_NUMBER || !isValidAsSequenceNumber(number)) {
         return INITIAL_SEQUENCE_NUMBER;
      }
      return number + 1;
   }

   // Private methods

   void SequenceNumber::validateArguments(std::initializer_list<int32_t> values)
   {
      for (int32_t value : values) {
         if (!isValidAsSequenceNumber(value)) {
            throw std::invalid_argument(
                  "Sequence number argument " + std::to_string(value) + " is invalid, must be in range " +
                  std::to_string(INITIAL_SEQUENCE_NUMBER) + " to " + std::to_string(MAX_SEQUENCE_NUMBER));
         }
      }
   }

} // FedPro
